### PR TITLE
Implement tracing for network-mux.

### PR DESCRIPTION
**Pull request by karknu**

### CODE REVIEW

**Comment by karknu** [comment]
https://github.com/input-output-hk/cardano-node/pull/205 contains the `cardano-node` side of this change (it will be updated once this PR is merged).

**Comment by karknu** [comment]
> these are quite a number of new trace sites. how much performance does this cost?

I haven't measured the performance impact but it better be zero when they are not enabled :).
The three events that are enabled by default  (MuxTraceStateChange, MuxTraceCleanExit, and MuxTraceExceptionExit) are very low frequency and shouldn't decrease performance.

File: network-mux/src/Network/Mux.hs [hunk 1]
```diff
@@ -101,14 +105,25 @@ muxStart peerid app bearer = do
                , muxControl pmss ModeResponder
                , muxControl pmss ModeInitiator
                ]
-    mjobs <- sequence [ mpsJob cnt pmss ptcl
+    mjobs <- sequence [ (mpsJob cnt pmss ptcl)
```
[the remark this answers is not in the document]
Comment (karknu) [reply network-mux/src/Network/Mux.hs]: Will address.

File: network-mux/src/Network/Mux.hs [hunk 2]
```diff
@@ -101,14 +105,25 @@ muxStart peerid app bearer = do
...
                       | ptcl <- [minBound..maxBound] ]
 
     mask $ \unmask -> do
-      as <- traverse (async . unmask) (jobs ++ concat mjobs)
-      muxBearerSetState bearer Mature
-      unmask (void $ waitAnyCancel as)
-      muxBearerSetState bearer Dead
+      as <- traverse (async . unmask) (jobs ++ (map fst $ concat mjobs))
```
[the remark this answers is not in the document]
Comment (karknu) [reply network-mux/src/Network/Mux.hs]: Nice, will make the change.

File: network-mux/src/Network/Mux.hs [hunk 3]
```diff
@@ -101,14 +105,25 @@ muxStart peerid app bearer = do
...
+      let aidsAndNames = zip as $ ["Demuxer", "Muxer", "MuxControl Responder", "MuxControl Initiator"] ++
+                                  (map snd $ concat mjobs)
+
+      muxBearerSetState tracer bearer Mature
+      unmask (do
+        (fa, r_e) <- waitAnyCatchCancel as
+        let faName = maybe "Unknown Protocol" id (lookup fa aidsAndNames)
```
[the remark this answers is not in the document]
Comment (karknu) [reply network-mux/src/Network/Mux.hs]: Will address.

File: network-mux/src/Network/Mux.hs [hunk 4]
```diff
@@ -101,14 +105,25 @@ muxStart peerid app bearer = do
...
-      as <- traverse (async . unmask) (jobs ++ concat mjobs)
-      muxBearerSetState bearer Mature
-      unmask (void $ waitAnyCancel as)
-      muxBearerSetState bearer Dead
+      as <- traverse (async . unmask) (jobs ++ (map fst $ concat mjobs))
+      let aidsAndNames = zip as $ ["Demuxer", "Muxer", "MuxControl Responder", "MuxControl Initiator"] ++
+                                  (map snd $ concat mjobs)
+
+      muxBearerSetState tracer bearer Mature
+      unmask (do
```
[the remark this answers is not in the document]
Comment (karknu) [reply network-mux/src/Network/Mux.hs]: Will address.

File: network-mux/src/Network/Mux/Bearer/Socket.hs [hunk 5]
```diff
@@ -54,21 +56,26 @@ socketAsMuxBearer sd = do
     where
       readSocket :: (HasCallStack) => IO (Mx.MuxSDU ptcl, Time IO)
       readSocket = do
+          traceWith tracer $ Mx.MuxTraceRecvHeaderStart
           hbuf <- recvLen' True 8 []
           --say "read"
           --hexDump hbuf ""
```
[the remark this answers is not in the document]
Comment (karknu) [reply network-mux/src/Network/Mux/Bearer/Socket.hs]: I will remove the debug messages along with `hexDump` when I've implemented support for pcap tracing.

File: ouroboros-network/src/Ouroboros/Network/NodeToNode.hs [hunk 6]
```diff
@@ -178,31 +180,33 @@ connectTo =
...
   -> NodeToNodeVersionData
   -> (OuroborosApplication InitiatorApp peerid NodeToNodeProtocols IO BL.ByteString a b)
   -> Maybe Socket.AddrInfo
   -> Socket.AddrInfo
   -> IO ()
-connectTo_V1 handshakeTracer peeridFn versionData application localAddr remoteAddr =
+connectTo_V1 muxTracer handshakeTracer peeridFn versionData application localAddr remoteAddr =
     connectTo
-      handshakeTracer peeridFn
+      muxTracer handshakeTracer peeridFn
```
[the remark this answers is not in the document]
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/NodeToNode.hs]: Will address.

File: ouroboros-network/src/Ouroboros/Network/Socket.hs [hunk 7]
```diff
@@ -225,20 +229,23 @@ beginConnection
...
     -- ^ either accept or reject a connection.
     -> Server.BeginConnection addr Socket.Socket st ()
-beginConnection handshakeTracer encodeData decodeData acceptVersion fn addr st = do
+beginConnection muxTracer handshakeTracer encodeData decodeData acceptVersion fn addr st = do
     accept <- fn addr st
     case accept of
       AcceptConnection st' peerid versions -> pure $ Server.Accept st' $ \sd -> do
-        (bearer :: MuxBearer ptcl IO) <- Mx.socketAsMuxBearer sd
-        Mx.muxBearerSetState bearer Mx.Connected
+        let muxTracer' = (Mx.WithMuxBearer (show peerid) `contramap` muxTracer)
```
[the remark this answers is not in the document]
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/Socket.hs]: Will address.

File: network-mux/src/Network/Mux/Types.hs [hunk 8]
```diff
@@ -286,3 +288,55 @@ instance Exception MuxError where
           (show errorMsg)
           (prettyCallStack errorStack)
 
+data WithMuxBearer a = WithMuxBearer {
```
[the remark this answers is not in the document]
Comment (karknu) [reply network-mux/src/Network/Mux/Types.hs]: Will address.

File: network-mux/src/Network/Mux/Types.hs [hunk 9]
```diff
@@ -286,3 +288,55 @@ instance Exception MuxError where
...
+      wmbTag   :: !String
+    , wmbEvent :: !a
+}
+
+instance Show a => Show (WithMuxBearer a) where
+    show WithMuxBearer {wmbTag, wmbEvent} = printf "Mux %s %s" wmbTag (show wmbEvent)
+
+data MuxTrace ptcl =
```
[the remark this answers is not in the document]
Comment (karknu) [reply network-mux/src/Network/Mux/Types.hs]: Will address

**Comment by coot** [comment]
the cost is non-zero (due to upstream calls to get current time and get thread id); @avieth has an idea how to make it noop when disabled. It might be good to measure the impact.

**Comment by karknu** [comment]
> the cost is non-zero (due to upstream calls to get current time and get thread id); @avieth has an idea how to make it noop when disabled. It might be good to measure the impact.

I thought that changes was implemented already. I'll measure the performance cost.

**Comment by karknu** [comment]
I've tested the performance cost of this change by running the mux `prop_mux_snd_recv` test case. That test runs in IO and is 100% cpu bound. The test was executed five times with 1000 quick-check tests each iteration and computed the average below.

|Setup                 | Test Time|
| --------------------:| --------:|
| mux with nullTracer  | 9.33s    |
| mux in `master`      | 8.79s     |

When actually printing the trace messages with showTracing it took 330s to run (I only did this once).

The test results corresponds to about a 6% increase in execution time. I think that the increased CPU cost is out weighed by the ability to troubleshoot node issues without a pcap. But @avieth 's tracing improvement would be very useful.

**Comment by coot** [comment]
That's not too bad; I am not sure if we measured the right thing though.  The cost that @CodiePP had in mind was due to the `LogObject`.  Did you try with a tracer:

```
debugNullTracer = contramapM (\a -> (,,a) <$> getCurrentTime <*> myThreadId) nullTracer
```

**Comment by karknu** [comment]
No, I only used a plain `nullTracer`. Will rerun with your tracer.

**Comment by karknu** [comment]
Here is an updated table.

|Setup                 | Test Time|
| --------------------:| --------:|
| mux with nullTracer  | 9.28s    |
| mux with debugNullTracer | 10.44s|
| mux in `master`      | 9.53s


So the increase is about 10% which I still find acceptable.

**Comment by coot** [comment]
If the performance will be too bad, we can use `Maybe (Tracer m ...)` and `traverse` over it.

File: network-mux/src/Network/Mux.hs [hunk 10]
```diff
@@ -126,34 +141,38 @@ muxStart peerid app bearer = do
...
 
         return $ case app of
-          MuxInitiatorApplication initiator -> [ initiator peerid mpdId initiatorChannel >> mpsJobExit cnt ]
-          MuxResponderApplication responder -> [ responder peerid mpdId responderChannel >> mpsJobExit cnt ]
-          MuxInitiatorAndResponderApplication initiator responder
-                                            -> [ initiator peerid mpdId initiatorChannel >> mpsJobExit cnt
-                                               , responder peerid mpdId responderChannel >> mpsJobExit cnt
-                                               ]
+          MuxInitiatorApplication initiator ->
+            [ (initiator peerid mpdId initiatorChannel >> mpsJobExit cnt , show mpdId ++ " Initiator")]
```
Comment (coot) [inline network-mux/src/Network/Mux.hs]: these lines might exceed now 80 chars

File: network-mux/src/Network/Mux.hs [hunk 11]
```diff
@@ -210,23 +230,34 @@ muxChannel pmss mid md cnt = do
...
 muxBearerSetState :: (MonadSTM m, Ord ptcl, Enum ptcl, Bounded ptcl)
-                  => MuxBearer ptcl m
+                  => Tracer m (MuxTrace ptcl)
+                  -> MuxBearer ptcl m
                   -> MuxBearerState
                   -> m ()
-muxBearerSetState bearer newState = atomically $ writeTVar (state bearer) newState
+muxBearerSetState tracer bearer newState = do
+    oldState <- atomically $ do
+        -- If MonadSTM had swapTVar we could use it here.
```
Comment (coot) [inline network-mux/src/Network/Mux.hs]: Maybe turn this into a todo, I created an issue for it: https://github.com/input-output-hk/ouroboros-network/issues/1063

File: network-mux/src/Network/Mux.hs (hunk 10 above)
Comment (karknu) [reply network-mux/src/Network/Mux.hs]: The line limit is 100 chars, but yes exceeds it. Will change.

File: network-mux/src/Network/Mux.hs (hunk 11 above)
Comment (karknu) [reply network-mux/src/Network/Mux.hs]: Will address.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 12]
```diff
@@ -75,7 +76,7 @@ import           Ouroboros.Storage.LedgerDB.InMemory (ledgerDbDefaultParams)
 --
 -- This function runs forever unless an exception is thrown.
 run
-  :: forall blk peer. (RunNode blk, Ord peer)
+  :: forall blk peer. (RunNode blk, Ord peer, Show peer)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: Why do we need this `Show peer`? Typically, using a `Tracer` avoids these constraints, as they only need to be satisfied by the `Tracer`, not by every use site of the it.

Comment (karknu) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: It is used when constructing an identifying tag for the Mux bearer.

Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: @karknu I found it strange too, maybe we can address this in the future.

Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: I opened an issue for this: https://github.com/input-output-hk/ouroboros-network/issues/1120

**Pull request merged**